# Selective sweep analysis in *Arabis nemorensis* and *Arabis sagittata*

This repository contains scripts, data files, and plots related to the identification of selective sweeps in *Arabis nemorensis* and *Arabis sagittata*. The analysis involves simulating demographic histories, detecting selective sweeps using OmegaPlus, and investigating the overlap between sweep regions and quantitative trait loci (QTLs).

## Files and descriptions

### plots/

#### `plot_overlap_QTLs_sweeps.R`

This R script visualizes the overlap between selective sweep regions identified by OmegaPlus and QTLs in *Arabis nemorensis* and *Arabis sagittata*.

### scripts/

#### `Arabis_divergent_model_singlepop_ancAndrecentMigration.slim`

This SLiM script simulates the demographic history of *Arabis nemorensis* and *Arabis sagittata*, incorporating both ancient and recent gene flow events. 
The simulation is based on the demographic history inferred by Dittberner et al. (2022) and includes:

- Two pulses of interspecific gene flow between the species:
  - An ancient pulse directly after the species split.
  - A recent migration event between sympatric populations.
- Generation of 10,000 neutral datasets of 2 Mb each.
- Assumption of a fixed recombination rate for each simulated block (5.5e-8).

The simulated data are used to establish cut-off values for the omega statistics to minimize false positives due to demographic processes.

#### `run_OmegaPlus_treesfile_simulatedData.py`

This Python script runs OmegaPlus on simulated data generated by the SLiM script. It:

- Reads tree files from the simulated datasets.
- Runs OmegaPlus with specified parameters.
- Outputs omega statistics for each simulated dataset.

Calculates omega statistics for simulated data to build a distribution for threshold determination.

#### `run_OmegaPlus_vcf_files_realData.py`

This Python script runs OmegaPlus on the Arabis data. It:

- Reads filtered VCF files.
- Runs OmegaPlus with specified parameters.
- Identifies selective sweep regions in the real dataset.

#### `read_OmegaPlus_vcf_files.py`

This Python script processes the output files from the analysis with OmegaPlus. 

### sweep_data_files/

This directory contains the output data files from OmegaPlus.

## Analysis Workflow

1. **Simulation of Demographic History:**
   - Use `Arabis_divergent_model_singlepop_ancAndrecentMigration.slim` to simulate the demographic history and generate neutral datasets under the model of ancient and recent migration.

2. **Selective Sweep Detection on Simulated Data:**
   - Run `run_OmegaPlus_treesfile_simulatedData.py` to calculate omega statistics on the simulated datasets.
   - Extract the maximum omega values from each dataset to build a distribution for threshold determination.

3. **Threshold Determination:**
   - Use the distribution of maximum omega values from the simulations to determine the 99th percentile threshold.

4. **Selective Sweep Detection on Real Data:**
   - Run `run_OmegaPlus_vcf_files_realData.py` with optimized parameters to detect selective sweeps in the real genomic data.
   - Process the output files of OmegaPlus using `read_OmegaPlus_vcf_files.py`.

5. **Overlap Analysis with QTLs:**
   - Use `plot_overlap_QTLs_sweeps.R` to visualize the selective sweep regions and their overlap with QTLs.
   - Identify overlapping regions where selective sweeps and QTL peaks co-localize within a 200 kb window.
   - Focus on the 10% quantile regions around the QTL peak positions for increased specificity.

6. **Visualization and Interpretation:**
   - Generate plots to illustrate the likelihood profiles of selective sweeps and their association with phenotypic traits.
   - Interpret the results to pinpoint candidate genomic regions under selection.
